Patrick: light wallet

thehogfather opened the pull request:
* Introduced Wallet Interface to capture the domain model of wallet along with two implementations of the interface for a local and remote wallet. These represent the Full Node and the Light Wallet implementations respectively. Also introduces a function for retrieving an instance of the active wallet based on the chain and the current wallet mode.
* Includes refactoring of stores that used to make api calls to the wallet domain object - now they depend on the networkStore for information on the active chain and the grab an instance of the wallet domain object and interact with the blockchain using its APIs.
* Includes menu in settings to allow switching between wallet modes.
* Includes simple screen in the onboarding flow to choose between a light wallet and a full node.
* Includes updates in the sidebar menu to disable items that are not relevant in the light wallet.
* Includes modal that shows prerequisites if an attempt is made to run wallet in full mode without meeting requirements

goldylucks committed:
add pagination to TxHistory

- add react-table
- delete SingleTxDelta
- add reusable Component ZpIoLink
- CopyableTableCell: refcator with ZpIoLink, extend to support ReactTable

thehogfather committed:
Patrick: WIP light wallet

* Introduced Wallet Interface to capture the domain model of wallet along with two implementations of the interface for a local and remote wallet. These represent the Full Node and the Light Wallet implementations respectively. Also introduces a function for retrieving an instance of the active wallet based on the chain and the current wallet mode.
* Includes refactoring of stores that used to make api calls to the wallet domain object - now they depend on the networkStore for information on the active chain and the grab an instance of the wallet domain object and interact with the blockchain using its APIs.
* Includes menu in settings to allow switching between wallet modes.
* Includes simple screen in the onboarding flow to choose between a light wallet and a full node.
* Includes updates in the sidebar menu to disable items that are not relevant in the light wallet.
* Includes modal that shows prerequisites if an attempt is made to run wallet in full mode without meeting requirements

thehogfather committed:
Patrick: WIP light wallet

* Introduced Wallet Interface to capture the domain model of wallet along with two implementations of the interface for a local and remote wallet. These represent the Full Node and the Light Wallet implementations respectively. Also introduces a function for retrieving an instance of the active wallet based on the chain and the current wallet mode.
* Includes refactoring of stores that used to make api calls to the wallet domain object - now they depend on the networkStore for information on the active chain and the grab an instance of the wallet domain object and interact with the blockchain using its APIs.
* Includes menu in settings to allow switching between wallet modes.
* Includes simple screen in the onboarding flow to choose between a light wallet and a full node.
* Includes updates in the sidebar menu to disable items that are not relevant in the light wallet.
* Includes modal that shows prerequisites if an attempt is made to run wallet in full mode without meeting requirements

Cremba commented:
1. it seems that the zen-node is still running or is not properly closed when it changes mode
2. let's add a back button in the onboarding "set up your wallet page"
3. the remote wallet is missing some functionalities

thehogfather commented:
> 1. it seems that the zen-node is still running or is not properly closed when it changes mode
It would be really helpful to figure out when this happens. Are you able to consistently reproduce?
> 2. let's add a back button in the onboarding "set up your wallet page"
Makes sense. 
> 3. the remote wallet is missing some functionalities
Could you be more specific please?

Cremba commented:
Sure I will check where the issue is 👍 
About the functionalities I tryed to send and execute contract which both failed

thehogfather committed:
Shutdown zen node before starting - to prevent errors.

Adds button to get back to choosing wallet mode screen

goldylucks commented on app/pages/Settings/Settings.js:
```diff
@@ -133,7 +137,7 @@ class Settings extends Component<Props, State> {
                 placeholder="Confirm new password"
                 className="input-group-field"
-                onPaste={disablePaste}
+                onPaste={undefined}
```
Why undefined and not remove this?

goldylucks replied:
resolved on last commit, closing this

goldylucks commented on app/pages/Settings/Settings.js:
```diff
@@ -173,17 +177,29 @@ class Settings extends Component<Props, State> {
-        <Flexbox flexDirection="column" className="description">
-          <h2 className="description-title">Wipe blockchain</h2>
-          <p>Wipe your local copy of the blockchain, with an option to wipe your wallet as well</p>
+    return this.props.walletModeStore.isFullNode() ?
```
I prefer to break it to "if/return", to be more aligned with the rest of the codebase:

```
if (this.props.walletModeStore.isFullNode()) {
  return (...)
}
return (...)
```

Thoughts?

goldylucks commented on app/pages/Settings/Settings.js:
```diff
@@ -310,6 +327,26 @@ class Settings extends Component<Props, State> {
+
+  renderWalletMode() {
+    const { walletModeStore } = this.props
+    const [currentWalletMode, otherWalletMode] = walletModeStore.modes
```
```suggestion
    const [currentWalletMode, otherWalletMode] = this.props.walletModeStore.modes
```

One liner instead of two?

goldylucks reviewed (commented):
Nice work man!

like I said in [trello](https://trello.com/c/hKWnfbx6/420-desktop-wallet-light-client-compatibility), refreshing the wallet takes me through onboarding again. Once that's done I'll continue the review.

YOU ROCK!

thehogfather commented on app/ZenNode.js:
```diff
@@ -94,6 +102,11 @@ class ZenNode {
 
+  onShutdownZenNode = (event, args) => {
+    this.config = { ...this.config, ...args }
+    this.shutdown()
```
need the ability to call shutdown on the node from the app. Using `this.node.kill` from `main.dev.js` breaks the law of demeter

goldylucks replied:
OK, agreeing to disagree here, not too concerning anyway, more of personal taste. Resolving this comment :)

thehogfather commented on app/config/server-address.js:
```diff
@@ -2,23 +2,24 @@
 const MAIN_NET_PORT = '11567'
 const LOCALHOST = 'http://127.0.0.1'
 
-let chain = getInitialChain()
```
I tested a few - but would be great if you can enumerate the concerning scenarios. I'll even write automated tests for them. The whole functionality of consistently switching between chains and wallet mode hinges on the db working. If it doesn't work, then the feature cannot be reliable.

goldylucks replied:
Don't know if it's related: I imported a wallet on testnet, then switched to mainnet. I was taken to onboarding process. Then I refreshed the app, and it prompts me for a password, even tho I don't have a wallet in the mainnet.

I'm guessing the check for existing wallet doesn't differentiate between chains now? If memory serves it used to.

thehogfather replied:
The check should differentiate. I'll have a look.

thehogfather committed:
Shutdown zennode if it is running on refresh

remove noop prop

thehogfather committed:
Persist network chain in db

thehogfather committed:
Added static types to track mapping between network names used in the
node, app and api
* only show mining in full node
* add cursor and hover effect to wallet choice

thehogfather committed:
Added static types to track mapping between network names used in the
node, app and api
* only show mining in full node
* add cursor and hover effect to wallet choice

goldylucks commented on app/main.dev.js:
```diff
@@ -12,15 +12,15 @@
 import db from './services/db'
 import MainProcessErrorReporter from './utils/errorReporting/MainProcessErrorReporter'
-import prereqCheck from './utils/prereqCheck'
+import prereqCheck, { PREREQUISITES_CHECK_FAILED } from './utils/prereqCheck'
```
Resolved on last commit, closing this conversation

goldylucks commented on package.json:
```diff
@@ -211,6 +211,7 @@
     "mobx-react": "^5.0.0",
     "moment": "^2.22.0",
     "normalize.css": "^8.0.0",
+    "npm": "^6.4.1",
```
Do we need npm intalled locally for something?

thehogfather replied:
no! nice catch - i probably ran npm update without the `-g` flag 🙈

goldylucks commented on app/stores/walletModeStore.js:
```diff
@@ -0,0 +1,33 @@
+      return this._mode === 'Full'
+    }
+
+    get modes() {
```
I prefer to use `mode` and `otherMode`, like we do in `networkStore` for `chain` and `otherChain`.

Reasons being:
1. currently, consumer of `modes` relies on unexplicit order of the modes. I prefer to explicitly see `mode`/`otherMode`
2. Scanning the `modes` getter requires additional investigation to know what it is used for (that's what I've just done - in order to understand how it's used, I've run a search across the project).
3. To be consistent with the codebase (same mechanism/logic as chain/otherChain)

Thoughts?

thehogfather replied:
The idea was to use a tuple type - but I can see how it adds some cognitive overhead - particularly with point 3. I'll update

thehogfather committed:
addressing code review comments

thehogfather committed:
addressing code review comments

thehogfather committed:
Add icons for wallet mode choice screen

* disable upload and run contract actions for light wallet

thehogfather committed:
Restart node when IPC signal is fired

* now works for wiping chain and toggling mining

goldylucks committed:
(TxHistory) change infinite scroll to pagination

- add react-table npm package
- delete Component SingleTxDelta
- add Component ZpIoLink
- add Component Dropdown
- CopyableTableCell: refcator with ZpIoLink, extend to support ReactTable

Cremba committed:
fix zp.io link

Cremba committed:
improved logic for changing chain

thehogfather committed:
Add wallet mode to netbottombar

* add zenjs version to sidebar

goldylucks committed:
fix lint

Cremba committed:
updated test in txStore

thehogfather committed:
updating package-lock.json

thehogfather committed:
Back button from wallet mode choice should send user to previous page
not beginning of previous section.

Cremba committed:
required UI fixes

thehogfather committed:
implement getTransactionsCount for web and local wallets

* add explicit buttons for choosing wallet modes

thehogfather committed:
Reuse same seed for both remote and local wallet when switching modes

* Adds a page to signal that a switch is in progress.

thehogfather committed:
Merge branch 'master' into thehogfather/lightwallet-r1

thehogfather committed:
removed extra sidebar menu

fix failing zennode test
